src/Microsoft.Azure.WebJobs.Host.Storage/BlobStorageConcurrencyStatusRepository.cs

```diff
@@ -0,0 +1,110 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.Storage;
+using Microsoft.Azure.Storage.Blob;
+using Microsoft.Azure.WebJobs.Host.Executors;
+using Microsoft.Azure.WebJobs.Host.Scale;
+using Microsoft.Azure.WebJobs.Logging;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json;
+
+namespace Microsoft.Azure.WebJobs.Host
+{
+    internal class BlobStorageConcurrencyStatusRepository : IConcurrencyStatusRepository
+    {
+        private const string HostContainerName = "azure-webjobs-hosts";
+        private readonly IHostIdProvider _hostIdProvider;
+        private readonly IConfiguration _configuration;
+        private readonly ILogger _logger;
+        private CloudBlobContainer _blobContainer;
+
+        public BlobStorageConcurrencyStatusRepository(IConfiguration configuration, IHostIdProvider hostIdProvider, ILoggerFactory loggerFactory)
+        {
+            _configuration = configuration;
+            _hostIdProvider = hostIdProvider;
+            _logger = loggerFactory.CreateLogger(LogCategories.Scale);
+        }
+
+        internal CloudBlobContainer Container
+        {
+            get
+            {
+                if (_blobContainer == null)
+                {
+                    string storageConnectionString = _configuration.GetWebJobsConnectionString(ConnectionStringNames.Storage);
+                    _blobContainer = CreateBlobContainer(storageConnectionString);
+                }
+                return _blobContainer;
+            }
+        }
+
+        public async Task<ConcurrencyStatusSnapshot> ReadAsync()
+        {
+            string blobPath = await GetBlobPathAsync();
+            try
+            {
+                CloudBlockBlob blob = Container.GetBlockBlobReference(blobPath);
+                if (await blob.ExistsAsync())
+                {
+                    string content = await blob.DownloadTextAsync();
+                    if (!string.IsNullOrEmpty(content))
+                    {
+                        var result = JsonConvert.DeserializeObject<ConcurrencyStatusSnapshot>(content);
+                        return result;
+                    }
+                }
+            }
+            catch (Exception e)
+            {
+                _logger.LogError(e, $"Error reading snapshot blob {blobPath}");
+                throw e;
+            }
+
+            return null;
```

Comment:
Consider introducing C# 8 nullable reference types for new code. That will make it easier for us to ensure we don't have sneaky issues where the caller of this method forgets to check for a potentially null result.

Reply:
I've set the language version for the WebJobs.Host and WebJobs.Host.Storage projects to latest (`<LangVersion>latest</LangVersion>` in proj file) and I've changed ReadAsync so the return type is annotated as nullable. I'm unable to enable the nullable context globally for these projects because it generates too many warnings outside of the scope of this PR. I have however annotated the source files that are the primary consumers of this API with `#nullable enable` so the compiler will generate warnings for them at least.